### PR TITLE
More completion improvements for PSCI

### PR DESCRIPTION
**Pull request by vkorablin**
Complete type names for `:k`, complete data constructors where identifiers are expected. Also only complete unqualified names when the module that contains them is imported. (Otherwise they're unavailable!)

I had to drill a little deeper into the data structures that represent modules; not all of the resulting stuff is what I'd consider 'pretty', so feedback is very welcome!

### CODE REVIEW

**Comment by paf31** [comment]
This looks good, but I'll merge it into my local copy and test first.

File: src/Language/PureScript/AST/Declarations.hs [hunk 1]
```diff
@@ -77,6 +81,7 @@ exportedDctors (Module _ decls exps) ident =
   where
   dctors = concatMap getDctors decls
   getDctors (DataDeclaration _ _ _ ctors) = map fst ctors
```
Comment (paf31) [inline src/Language/PureScript/AST/Declarations.hs]: Probably need to handle data binding groups here too.

File: psci/Main.hs [hunk 2]
```diff
@@ -238,30 +240,60 @@ completion = completeWordWithPrev Nothing " \t\n\r" findCompletions
...
+            getTypeName (P.TypeSynonymDeclaration name _ _) = Just name
+            getTypeName (P.DataDeclaration _ name _ _) = Just name
+            getTypeName (P.PositionedDeclaration _ d) = getTypeName d
+            getTypeName _ = Nothing
+
+  identNames :: P.Module -> [N.Ident]
+  identNames (P.Module _ ds exports) = nub [ ident | ident <- mapMaybe (getDeclName exports) ds ]
+    where getDeclName :: Maybe [P.DeclarationRef] -> P.Declaration -> Maybe P.Ident
+          getDeclName exts decl@(P.ValueDeclaration ident _ _ _)  | P.isExported exts decl = Just ident
+          getDeclName exts decl@(P.ExternDeclaration _ ident _ _) | P.isExported exts decl = Just ident
```
Comment (paf31) [inline psci/Main.hs]: Probably need to handle binding groups here too.

**Comment by paf31** [comment]
Just tested, and everything looks great, but for some reason `:t f` doesn't autocomplete to `:t flip` for my any more. Any idea why?

**Comment by vkorablin** [comment]
Thanks for the feedback, I'll take a look.

**Comment by vkorablin** [comment]
Regarding breakage: `Language.PureScript.AST.Declarations.isExported` didn't handle a certain case. I'll take care of the binding groups tomorrow.

**Comment by paf31** [comment]
We can fix binding groups separately if you like. I'd like to make a release of the compiler today, and it would be nice to have this included, if the `flip` issue is a simple fix.

**Comment by vkorablin** [comment]
Sure, go ahead. c542b4ff52695a753d95d7a50c46223995bdebd8 has the fix for that. Heads up: including `Language.PureScript.Sugar.BindingGroups` (for `collapseBindingGroups`/`collapseBindingGroupsModule`) from `Language.PureScript.AST.Declarations` results in an import loop.

**Comment by paf31** [comment]
:+1: Awesome, thanks!

I don't think we need to import `collapseBindingGroups`, just unwrap those declarations inside the binding group. We can deal with it separately though.

**Pull request merged**